subprojects/model-core/src/main/java/org/gradle/api/internal/provider/DefaultProviderFactory.java

```diff
@@ -210,7 +210,7 @@ public <T extends Credentials> Provider<T> credentials(Class<T> credentialsType,
 
     @Override
     public <A, B, R> Provider<R> zip(Provider<A> left, Provider<B> right, BiFunction<? super A, ? super B, ? extends R> combiner) {
-        return new BiProvider<>(left, right, combiner);
+        return left.zip(right, combiner);
```

Comment:
I think it's better to delegate here, since we don't do anything special anyway.
With delegation, any improvements to `.zip` (e.g. better `getType` inference) will automatically apply here.